pkg/controller/directvolumemigration/rsync.go

```diff
@@ -1651,3 +1678,591 @@ func (t *Task) deleteProgressReportingCRs(client k8sclient.Client) error {
+	}
+	return t.Owner.Spec.BackOffLimit
+}
+
+// runRsyncOperations creates pod requirements for Rsync pods for all PVCs present in the spec
+// runs Rsync operations for all PVCs concurrently, processes outputs of each operation
+// returns whether or not all operations are completed, whether any of the operation is failed, and a list of failure reasons
+func (t *Task) runRsyncOperations() (bool, bool, []string, error) {
+	var failureReasons []string
+	srcClient, podRequirements, err := t.getRsyncOperationsContext()
+	if err != nil {
+		return false, false, failureReasons, liberr.Wrap(err)
+	}
+	status, err := t.ensureRsyncOperations(srcClient, podRequirements)
+	if err != nil {
+		return false, false, failureReasons, liberr.Wrap(err)
+	}
+	// report progress of pods
+	t.reportRsyncPodProgress()
+	isComplete, anyFailed, failureReasons, err := t.processRsyncOperationStatus(status)
+	if err != nil {
+		return false, false, failureReasons, liberr.Wrap(err)
+	}
+	return isComplete, anyFailed, failureReasons, nil
+}
+
+// processRsyncOperationStatus processes status of Rsync operations by reading the status list
+// returns whether all operations are completed and whether any of the operation is failed
+func (t *Task) processRsyncOperationStatus(status rsyncClientOperationStatusList) (bool, bool, []string, error) {
+	isComplete, anyFailed, failureReasons := false, false, make([]string, 0)
+	if status.AllCompleted() {
+		isComplete = true
+		// we are done running rsync, we can move on
+		// need to check whether there are any permanent failures
+		if status.Failed() > 0 {
+			anyFailed = true
+			// attempt to categorize failures in any of the special failure categories we defined
+			failureReasons, err := t.reportAdvancedErrorHeuristics()
+			if err != nil {
+				return isComplete, anyFailed, failureReasons, liberr.Wrap(err)
+			}
+		}
+		return isComplete, anyFailed, failureReasons, nil
+	}
+	if status.AnyErrored() {
+		// check if we are seeing errors running any of the operation for over 5 minutes
+		// if yes, set a warning condition
+		runningCondition := t.Owner.Status.Conditions.FindCondition(Running)
+		if runningCondition != nil &&
+			time.Now().Add(time.Minute*-5).After(runningCondition.LastTransitionTime.Time) {
+			t.Owner.Status.SetCondition(migapi.Condition{
+				Category: Warn,
+				Type:     FailedCreatingRsyncPods,
+				Message:  "Repeated errors occurred when attempting to create one or more Rsync pods in the source cluster. Please check controller logs for details.",
+				Reason:   Failed,
+				Status:   True,
+			})
+		}
+	}
+	return isComplete, anyFailed, failureReasons, nil
+}
+
+// reportAdvancedErrorHeuristics processes DVMP CRs for all PVCs,
+// for all errored pods, attempts to determine whether the errors fall into any
+// of the special categories we can identify and reports them as conditions
+// returns reasons and error for reconcile decisions
+func (t *Task) reportAdvancedErrorHeuristics() ([]string, error) {
```

Comment:
these functions were previously called directly in the switch case for itinerary, moved them outside here to keep itinerary  logic easy and readable